### DISCUSSION TITLE
Issues in output inherited from base class 'CrystalPlasticityStressUpdateBase'

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#20593] [Q&A Modules: Solid mechanics] [answered]
Hi moose experts or users,

I recently developed a custom class studying creep behavior (e.g. strain vs. time, strain rate vs. time)in which the base class **CrystalPlasticityStressUpdateBase** is used. I checked some functions in the base class that has been overwritten in the child class and printed out some key variables. All look good. The problem is that all outputted strain_zz are 0 under a certain temperature and pressure in BCs. But if I used tensile load like '0.01*t', there will output non-zero values. This is very strange. I hope the simulation is performed using the former conditions. Could anyone share ideas and advice with me to fix it?

**Comment by sapitts** [comment]
Hi @xchengood, 
Have you been able to replicate this issue with one of the crystal plasticity classes (that inherits from `CrystalPlasticityStressUpdateBase`) in the MOOSE repository? If not, would you please create an input file that uses one of the classes checked into the MOOSE repository please? I'd recommend using `CrystalPlasticityKalidindiUpdate` and just the pressure BC to start

Having an input file that uses a checked-in class will allow us to narrow where the problem occurs. And having a MOOSE-repository based input file means it will be easier for me to look at your input file too.

Hope this helps,
Stephanie

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for your advice. I tried to use **CrystalPlasticityKalidindiUpdate** and **CrystalPlasticityHCPDislocationSlipBeyerleinUpdate** and both are set pressure as BCs.

For the **CrystalPlasticityKalidindiUpdate**, the input file is below.
[update_method_test-1.txt](https://github.com/idaholab/moose/files/8343326/update_method_test-1.txt)

The outputted total lagrangian strain (e_zz) and plastic deformation gradient (fp_zz) as a function of time at 1350K under 200MPa look good as shown below.
<img width="723" alt="Screen Shot 2022-03-24 at 10 51 30 AM" src="https://user-images.githubusercontent.com/74226455/159956753-ed56f89c-c482-4398-9eaf-2822150d1ec8.png">

For the **CrystalPlasticityHCPDislocationSlipBeyerleinUpdate**, the input file is below.
[update_method_hcp_no_substructure111.txt](https://github.com/idaholab/moose/files/8343323/update_method_hcp_no_substructure111.txt)

But the outputted total lagrangian strain (e_zz) and plastic deformation gradient (fp_zz) at 1300K under 100MPa look weird and are constant. The curves are shown below. I thought this is probably caused by not setting high enough temperature and tensile stress, So I try at 1600K and 200MPa. It still outputs constant e_zz and fp_zz.
<img width="722" alt="Screen Shot 2022-03-24 at 10 56 09 AM" src="https://user-images.githubusercontent.com/74226455/159957824-68ecb546-c671-44ce-9c5e-8fdca5c4d91d.png">


> Having an input file that uses a checked-in class will allow us to narrow where the problem occurs?

Sorry, could you share more about how to use a checked-in class with me in an input file?

Thank you.

**Reply by sapitts** [reply]
> Sorry, could you share more about how to use a checked-in class with me in an input file?

You've done this with the attached text files. Thanks for adding those! I'll take a look and get back to you next week

**Reply by xchengood** [reply]
Thank you.

**Reply by sapitts** [reply]
Hi @xchengood,
There are a couple of changes that needed to be made to the input file using the the ` CrystalPlasticityHCPDislocationSlipBeyerleinUpdate ` class to show variation in the strain and plastic deformation measures (in the zz component):

- Use of physical values for the substructure rate coefficients: Change the line 
  ` substructure_rate_coefficient_per_mode = '-355 -0.4' #artifical, non-physical values for testing purposes ` 
to reflect physical properties, such as 
  ` substructure_rate_coefficient_per_mode = '355 0.4' #from Capolungo et al MSEA (2009) `
as given in the regression test [update_method_hcp_aprismatic_capyramidal.i](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/hcp_single_crystal/update_method_hcp_aprismatic_capyramidal.i).
- Apply a higher stress value so as to overcome the slip system resistance: 
    In the provided input file, the resolved shear stress on the active slip systems was less than 3% of the slip system resistance. As a result, no dislocation evolution occurred in the simulation.

The simulation results from these input file changes do show the evolution of the total lagrangian strain (beyond the initial elastic response you showed above) and plastic deformation gradient. For quick reference, here are the two input files I used to investigate this issue:

- Modified input file: [userlist_hcp_corrected.txt](https://github.com/idaholab/moose/files/8401331/userlist_hcp_corrected.txt)
- Original input file, with additional outputs for the dislocation densities: 
[userlist_hcp_original_extra_outputs.txt](https://github.com/idaholab/moose/files/8401332/userlist_hcp_original_extra_outputs.txt)
 
I'll add figures to compare the results early next week. I hope the comparison input files will help in debugging your custom creep behavior class

**Reply by xchengood** [reply]
@sapitts. Thank you very much. That is very helpful. Let me run it and look at how the strain changes with time.

**Comment by sapitts** [comment, marked as answer]
Here are the plots comparing the original hcp input file ([userlist_hcp_original_extra_outputs.txt](https://github.com/idaholab/moose/files/8414138/userlist_hcp_original_extra_outputs.txt))
and the modified input file ([userlist_hcp_corrected_tension.txt](https://github.com/idaholab/moose/files/8414139/userlist_hcp_corrected_tension.txt)). Please note that I've corrected the modified input file from Friday to use a tensile pressure instead of the default compressive pressure.

The total lagranian strain and the plastic deformation gradient show elastic behavior for the original input and plastic deformation behavior for the modified input file (temperature constant at 1300K):
![Comparison_hcp_compression_uniaxial_compression_strain_zz](https://user-images.githubusercontent.com/9342949/161659583-b16ebcf7-72b0-4594-b7b7-d35dc0212680.png)   ![Comparison_hcp_compression_deformation_gradient_zz](https://user-images.githubusercontent.com/9342949/161659774-1451a133-eecc-40b6-8020-7889d4dcbb80.png)

This difference in behavior is primarily because the slip resistance (on the first prismatic slip system which is active in this loading orientation)  is an order of magnitude larger than the applied resolved shear stress on the same system:
![Comparison_hcp_compression_active_slip_system_resistance](https://user-images.githubusercontent.com/9342949/161661547-f29d0333-aafe-4c96-a4dd-2d0ca2de93a5.png)
![Comparison_hcp_compression_active_system_resolved_shear_stress](https://user-images.githubusercontent.com/9342949/161661580-9f3133dd-80ce-4afe-a86e-ca4890ece8ed.png)

This mismatch in the applied resolved shear stress and the slip system resistance is such that the forest dislocation density (on the same prismatic slip system) and substructure density (total quantity) can only evolve in the modified input file:
![Comparison_hcp_compression_forest_dislocation_evolution](https://user-images.githubusercontent.com/9342949/161661888-ca03437f-f89d-4dd8-8b5c-cc4a9c1e298a.png)
![Comparison_hcp_compression_substructure_evolution](https://user-images.githubusercontent.com/9342949/161661929-ec5100ae-9d4c-44d6-9ee3-6a631827f6c2.png)

**Reply by xchengood** [reply]
Hi @sapitts. Sorry for the late reply. Thank you very much for your detailed comparison information. I ran it and got the same results and had a careful look at the output variable values. It is really helpful for me to check my test file. 

Last one small question, what software do you use to draw those figures. They look brief but beautiful. It seems not to use Excel or origin.

**Reply by ngrilli** [reply]
@xchengood  should be matplotlib I guess

**Reply by sapitts** [reply]
Hi @xchengood, @ngrilli  is correct, I used matplotlib and pandas to generate those graphs. Here's the python script (which you will have to resave as a `*.py` file in order to run):
[plot_physical_quantities_comparison.txt](https://github.com/idaholab/moose/files/8497153/plot_physical_quantities_comparison.txt)

**Reply by xchengood** [reply]
@ngrilli and @sapitts, Thank you very much.
